src/classes/trackers.js

```diff
@@ -1,290 +1,498 @@
+        const siteDomain = this.tldts.parse(siteUrl).domain
+        const urlToCheckDomain = this.tldts.parse(urlToCheck).domain
+        if (!siteDomain || !urlToCheckDomain) {
+            return null
+        }
+        return {
+            siteUrl: siteUrl,
+            request: request,
+            siteDomain,
+            siteUrlSplit: this.utils.extractHostFromURL(siteUrl).split('.'),
+            urlToCheck: urlToCheck,
+            urlToCheckDomain,
+            urlToCheckSplit: this.utils.extractHostFromURL(urlToCheck).split('.')
+        }
+    }
 
-            return {
-                action,
-                reason,
-                firstParty,
-                redirectUrl,
-                matchedRule,
-                matchedRuleException,
-                tracker,
-                fullTrackerDomain,
-                fromCname
+    /**
+     * @param {string} url
+     * @returns {boolean}
+     */
+    isSpecialURL (url) {
+        let urlObj
+        try {
+            urlObj = new URL(url)
+        } catch {
+            // This really shouldn't happen but if it does, we'll just assume it's a special URL
+            return true
+        }
+        const specialProtocols = ['chrome-extension:', 'chrome-devtools:', 'chrome-search:', 'about:', 'file:', 'javascript:', 'data:']
+        if (urlObj) {
+            if (specialProtocols.includes(urlObj.protocol)) {
+                return true
             }
         }
+        return false
+    }
 
-        /*
-         * Pull subdomains off of the reqeust rule and look for a matching tracker object in our data
-         */
-        findTracker (requestData) {
-            const urlList = Array.from(requestData.urlToCheckSplit)
+    /**
+     * @param {string} urlToCheck
+     * @param {string} siteUrl
+     * @param {RequestExpression} request
+     * @returns {TrackerData | null}
+     */
+    getTrackerData (urlToCheck, siteUrl, request) {
+        if (!this.entityList || !this.trackerList) {
+            throw new Error('tried to detect trackers before rules were loaded')
+        }
 
-            while (urlList.length > 1) {
-                const trackerDomain = urlList.join('.')
-                urlList.shift()
+        if (this.isSpecialURL(urlToCheck)) {
+            return null
+        }
 
-                const matchedTracker = this.trackerList[trackerDomain]
-                if (matchedTracker) {
-                    return matchedTracker
+        let fromCname
+        let requestData = this.getRequestData(urlToCheck, siteUrl, request)
+        if (!requestData) {
+            return null
+        }
+        const requestOwner = this.findTrackerOwner(requestData.urlToCheckDomain)
+        const websiteOwner = this.findWebsiteOwner(requestData)
+
+        const firstParty = (requestOwner && websiteOwner) ? requestOwner === websiteOwner : requestData.siteDomain === requestData.urlToCheckDomain
+        const fullTrackerDomain = requestData.urlToCheckSplit.join('.')
```

Comment:
Nit: rename to fullRequestDomain, since we don't yet know if a tracker?

Reply:
Ignoring as this would change the output property which would require extension changes.